### ISSUE TITLE
Nearby users changesets times out

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#2396]
From https://trac.openstreetmap.org/ticket/4176

Visiting http://www.openstreetmap.org/browse/nearby (linked to from user profile page) times out, at least for me. I'm surprised by this, since the nearby users list works fine, and so does changesets from friends, so I don't know why changesets from nearby users causes the timeout. 

Perhaps something goes disastrously wrong with the query planner? Or perhaps we should build the query as a join rather than subselect? I haven't yet reproduced this locally.

**Comment by tomhughes** [comment]
This has always been troublesome but I thought I had fixed it by tiling the users some time ago?

**Comment by tomhughes** [comment]
Oh sorry I was talking about the list of nearby users on the user page - no idea what that is doing but it likely needs to include the user tile numbers in the filter if it isn't already.

**Comment by gravitystorm** [comment]
It does so already, see e.g. https://gist.github.com/gravitystorm/3bc58ed852d7b7ff6c3ff5f58e720c7a

**Comment by tomhughes** [comment]
Right, so the problem is the sort on changeset ID combined with the relatively low cardinality selection on the user ID of the changeset.

So the planner basically has two options, neither of them very good:

* Use the user index to find all the changesets for nearby users, then sort and limit that - wasteful because you have to find all the changsets for the user in order to then do the pagination.
* Use the changeset ID index to read the changesets in order, filtering on user ID until we have the number of results we want.

It is going (for my user at least) for the second option, but it will have to read through vast numbers of changesets by other people to try and find 20 by nearby users.

If there was an index on user_id+changeset_id then it might do better using that but given we want multiple users I'm not sure if it would be clever enough - it would either have to do a bunch of parallel index scans and merge them until it had enough results, or read 20 for user user then sort and discard the unwanted ones.

**Comment by pnorman** [comment]
Rebuilding the indexes might also help - everything in the users table tends to become bloated

**Comment by tomhughes** [comment]
It has no problem finding the nearby users - the slow thing is finding the changesets.
